### PR TITLE
Fixed out= variant of linalg.inv

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR modifies the behavior of the `linalg_inv_out` variant to match the description here https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch
With this PR result and input tensors must be on the same device and have the same "type kind".
It's allowed to pass out tensors with complex dtypes for float inputs.

Ref. https://github.com/pytorch/pytorch/issues/42666

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:713]: Instead of recursively calling the operation, would it be more readable to setup result_tmp either this way or as a copy of input, then check `copy_needed` before returning from this function and, if needed, perform the copy to the original result then?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:742]: Would you elaborate on why these are changing?

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebraUtils.h:350]: Nice helper.

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebraUtils.h:362]: This is true but a little misleading since we allow other casts, too. This comment should probably just refer to the use of the canCast() function.

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebraUtils.h:368]: Expected input's dtype to be (safely) castable to result's dtype.

We don't have a notion of "compatible" dtypes.

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebraUtils.h:378]: Same wording suggestion here.

**Comment by mruberry** [comment]
Hey @IvanYashchuk! This looks great, just a few inline comments/questions. If I'm unavailable when you'd like this merged ping @heitorschueroff and he'll help get it landed.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:713]: The idea is to avoid both copy and temporary tensors if possible, that's why checking `copy_needed` should happen before the actual computations and not right before the return.

I can remove the recursion.
Currently, we have a hierarchy of calls:

1. `linalg_inv` just creates out tensor and calls `linalg_inv_out`
2. `linalg_inv_out` just creates infos tensors and calls `linalg_inv_out_info`
(2) it's simple to add a variant that doesn't do `infos` synchronization and checking for error here
3. `linalg_inv_out_info` checks and prepares inputs for calling LAPACK/MAGMA/cuSOLVER
3.1 if `result` is not good for LAPACK allocate a temporary and call recursively `linalg_inv_out_info`
4. `_linalg_inv_out_helper` dispatch to LAPACK/MAGMA/cuSOLVER

So checking if the temporary tensor and copy is needed could happen in step 2, then it would need to be duplicated in the future `linalg_inv_out_without_info_check` function. I can add a separate function between 3 and 4 to avoid the recursive call.

Comment (mruberry) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:713]: Sounds like it would just complicate things more to be rid of the recursive call? Probably not worth it, then.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:742]: So that we don't do the resizing and filling with zeros later. Filling with zeros is needed because the underlying library calls not always do that leading to spurious exception throws due to non-zero content.
https://github.com/pytorch/pytorch/pull/50269#discussion_r556145441

**Comment by IvanYashchuk** [comment]
I updated the error message on typecasting. But this PR should wait until https://github.com/pytorch/pytorch/pull/51560 lands and then I'll rebase because there would be a merge conflict in helper functions.

**Comment by IvanYashchuk** [comment]
@mruberry I've changed the error message "compatible dtypes" -> "castable dtypes".

I think we can merge it now. The ROCm failure seems to be not related.

**Comment by heitorschueroff** [comment]
LGTM, @mruberry I can land this once you approve it.

**Review by mruberry — APPROVED** [review]
Awesome!

**Pull request closed without merging**
